pin-project-internal/src/pin_projectable/mod.rs

```diff
@@ -0,0 +1,255 @@
+fn proj_generics(generics: &Generics) -> Generics {
+    let mut generics = generics.clone();
+    generics.params.insert(0, syn::parse_quote!('__a));
+    generics
+}
+
+struct ImplDrop {
+    generics: Generics,
+    pinned_drop: Option<ItemFn>,
+}
+
+impl ImplDrop {
+    /// Parses attribute arguments.
+    fn new(generics: Generics, pinned_drop: Option<ItemFn>) -> Result<Self> {
+        Ok(ImplDrop { generics, pinned_drop })
+    }
+
+    fn build(self, ident: &Ident) -> TokenStream {
+        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
+
+        match self.pinned_drop {
+            Some(fn_) => {
+                let fn_name = fn_.ident.clone();
+                quote! {
+                    impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
+                        fn drop(&mut self) {
+                            // Declare the #[pinned_drop] function *inside* our drop function
+                            // This guarantees that it's impossible for any other user code
+                            // to call it
+                            #fn_
+                            // Safety - we're in 'drop', so we know that 'self' will
+                            // never move again
+                            let pinned_self = unsafe { ::core::pin::Pin::new_unchecked(self) };
+                            // 'pinned_drop' is a free function - if it were part of a trait impl,
+                            // it would be possible for user code to call it by directly invoking
+                            // the trait.
+                            // Therefore, we enforce a return type of '()' by explicitly
+                            // assigning it to a temporary.
+                            let _: () = #fn_name(pinned_self);
+                        }
+                    }
+                }
+            }
+            None => {
+                quote! {
+                    impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
+                        fn drop(&mut self) {
+                            // Do nothing. The precense of this Drop
+                            // impl ensures that the user can't provide one of their own
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+// =================================================================================================
+// conditional Unpin implementation
+
+struct ImplUnpin {
+    generics: Generics,
+    auto: bool,
+}
+
+impl ImplUnpin {
+    /// Parses attribute arguments.
+    fn new(args: TokenStream, generics: &Generics) -> Result<Self> {
+        let mut generics = generics.clone();
+        generics.make_where_clause();
+
+        match &*args.to_string() {
+            "" => Ok(Self { generics: generics.clone(), auto: true }),
+            "unsafe_Unpin" => Ok(Self { generics: generics.clone(), auto: false }),
```

Comment:
Another idea for this would be to require it be an `unsafe impl Unpin` inside a `pin_project!` block, something like

```rust
//! WARNING: Unsound code

pin_project! {
    #[pin_projectable]
    pub struct Blah<T> {
        field: T,
    }

    unsafe impl<T> Unpin for Blah<T> {}
}
```

I don't know how this interacts with tools like `cargo-geiger`, since the `unsafe` keyword won't make it into the expanded code, but it at least gets it there much more obviously when checking via `rg`.